File: aiwendy/apps/api/routers/ollama.py
"""API routes for Ollama model management."""

from fastapi import APIRouter, HTTPException, Depends
from fastapi.responses import StreamingResponse
from typing import List, Dict, Any
from pydantic import BaseModel
import json

from core.auth import get_current_user
from core.logging import get_logger
from domain.user.models import User
from infrastructure.llm.ollama_provider import OllamaProvider

logger = get_logger(__name__)
router = APIRouter()


class OllamaModelInfo(BaseModel):
    """Model information."""
    name: str
    modified_at: str
    size: int
    digest: str


class OllamaHealthResponse(BaseModel):
    """Health check response."""
    healthy: bool
    message: str


class PullModelRequest(BaseModel):
    """Request to pull a model."""
    model_name: str


class ListModelsResponse(BaseModel):
    """Response for list models."""
    models: List[str]
    available: bool


@router.get("/health", response_model=OllamaHealthResponse)
async def check_ollama_health():
    """Check if Ollama service is running."""
    provider = OllamaProvider()
    is_healthy = await provider.check_health()

    return OllamaHealthResponse(
        healthy=is_healthy,
        message="Ollama service is running" if is_healthy else "Ollama service is not available"
    )


@router.get("/models", response_model=ListModelsResponse)
async def list_models():
    """List available models in Ollama."""
    provider = OllamaProvider()

    # Check if service is running
    is_healthy = await provider.check_health()
    if not is_healthy:
        return ListModelsResponse(
            models=[],
            available=False
        )

    models = await provider.list_models()
    return ListModelsResponse(
        models=models,
        available=True
    )


@router.post("/models/pull")
async def pull_model(
    request: PullModelRequest,
    current_user: User = Depends(get_current_user)
):
    """Pull a model from Ollama registry."""
    provider = OllamaProvider()

    # Check if service is running
    is_healthy = await provider.check_health()
    if not is_healthy:
        raise HTTPException(
            status_code=503,
            detail="Ollama service is not available"
        )

    async def generate():
        """Generate SSE stream for model pulling progress."""
        async for status in provider.pull_model(request.model_name):
            # Format as Server-Sent Events
            event_data = json.dumps({"status": status})
            yield f"data: {event_data}\n\n"

        # Send final message
        yield f"data: {json.dumps({'status': 'Complete', 'done': True})}\n\n"

    return StreamingResponse(
        generate(),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "Connection": "keep-alive",
            "X-Accel-Buffering": "no",
        }
    )


@router.get("/recommended-models")
async def get_recommended_models():
    """Get list of recommended models for trading psychology coaching."""
    return {
        "models": [
            {
                "name": "llama3.2:latest",
                "description": "Latest Llama 3.2 model - Fast and capable",
                "size": "3.2GB",
                "recommended": True,
                "use_case": "General coaching and analysis"
            },
            {
                "name": "mistral:latest",
                "description": "Mistral 7B - Efficient and powerful",
                "size": "4.1GB",
                "recommended": True,
                "use_case": "Technical analysis and pattern recognition"
            },
            {
                "name": "phi3:medium",
                "description": "Microsoft Phi-3 Medium - Compact but capable",
                "size": "7.9GB",
                "recommended": False,
                "use_case": "Quick responses and real-time coaching"
            },
            {
                "name": "gemma2:9b",
                "description": "Google Gemma 2 9B - Advanced reasoning",
                "size": "5.4GB",
                "recommended": True,
                "use_case": "Deep analysis and improvement plans"
            },
            {
                "name": "qwen2.5:7b",
                "description": "Qwen 2.5 7B - Multilingual support",
                "size": "4.4GB",
                "recommended": False,
                "use_case": "Multilingual coaching"
            },
            {
                "name": "nomic-embed-text",
                "description": "Embedding model for semantic search",
                "size": "274MB",
                "recommended": True,
                "use_case": "Journal search and similarity"
            }
        ]
    }


@router.post("/test-chat")
async def test_ollama_chat(
    model: str,
    message: str,
    current_user: User = Depends(get_current_user)
):
    """Test chat with a specific Ollama model."""
    provider = OllamaProvider()

    # Check if service is running
    is_healthy = await provider.check_health()
    if not is_healthy:
        raise HTTPException(
            status_code=503,
            detail="Ollama service is not available"
        )

    # Check if model is available
    models = await provider.list_models()
    if model not in models:
        raise HTTPException(
            status_code=404,
            detail=f"Model {model} is not available. Please pull it first."
        )

    # Create test messages
    from infrastructure.llm.base import Message, LLMConfig

    messages = [
        Message(
            role="system",
            content="You are a trading psychology coach. Provide helpful and concise responses."
        ),
        Message(
            role="user",
            content=message
        )
    ]

    config = LLMConfig(
        model=model,
        temperature=0.7,
        max_tokens=500,
        stream=False
    )

    # Get response
    response = await provider.chat(messages, config)

    return {
        "model": model,
        "message": message,
        "response": response
    }

File: keeltrader/apps/web/lib/api/ollama.ts
/**
 * Ollama API client for local model management
 */

import { apiJson, apiStream } from '@/lib/api/client'

export interface OllamaHealthResponse {
  healthy: boolean
  message: string
}

export interface OllamaModel {
  name: string
  modified_at?: string
  size?: number
  digest?: string
}

export interface ListModelsResponse {
  models: string[]
  available: boolean
}

export interface RecommendedModel {
  name: string
  description: string
  size: string
  recommended: boolean
  use_case: string
}

export interface PullProgress {
  status: string
  done?: boolean
}

export interface TestChatResponse {
  model: string
  message: string
  response: string
}

class OllamaApi {
  /**
   * Check if Ollama service is running
   */
  async checkHealth(): Promise<OllamaHealthResponse> {
    return apiJson<OllamaHealthResponse>('/ollama/health')
  }

  /**
   * List available models in Ollama
   */
  async listModels(): Promise<ListModelsResponse> {
    return apiJson<ListModelsResponse>('/ollama/models')
  }

  /**
   * Get recommended models for trading psychology coaching
   */
  async getRecommendedModels(): Promise<{ models: RecommendedModel[] }> {
    return apiJson<{ models: RecommendedModel[] }>('/ollama/recommended-models')
  }

  /**
   * Pull a model from Ollama registry
   * @param modelName - Name of the model to pull
   * @param onProgress - Callback for progress updates
   */
  async pullModel(
    modelName: string,
    onProgress?: (progress: PullProgress) => void
  ): Promise<void> {
    const response = await apiStream('/ollama/models/pull', {
      method: 'POST',
      body: { model_name: modelName },
    })

    // Handle Server-Sent Events stream
    const reader = response.body?.getReader()
    const decoder = new TextDecoder()

    if (!reader) {
      throw new Error('No response body')
    }

    let buffer = ''

    while (true) {
      const { done, value } = await reader.read()

      if (done) {
        break
      }

      buffer += decoder.decode(value, { stream: true })

      // Process complete messages
      const lines = buffer.split('\\n')
      buffer = lines.pop() || ''

      for (const line of lines) {
        if (line.startsWith('data: ')) {
          try {
            const data = JSON.parse(line.slice(6))
            if (onProgress) {
              onProgress(data)
            }
            if (data.done) {
              return
            }
          } catch (e) {
            console.error('Failed to parse SSE data:', e)
          }
        }
      }
    }
  }

  /**
   * Test chat with a specific Ollama model
   */
  async testChat(model: string, message: string): Promise<TestChatResponse> {
    return apiJson<TestChatResponse>('/ollama/test-chat', {
      method: 'POST',
      body: { model, message },
    })
  }
}

export const ollamaApi = new OllamaApi()
